src/api/app/controllers/staging/workflows_controller.rb

```diff
@@ -20,22 +23,46 @@ def create
   end
 
   def destroy
-    staging_workflow = @project.staging
-    unless staging_workflow
+    authorize @staging_workflow
+
+    @staging_workflow.destroy!
+  end
+
+  def merge_staging_projects
+    authorize @staging_workflow, :update?
+
+    projects_to_merge = @staging_workflow.staging_projects.where(name: @staging_projects)
+    missing_projects = @staging_projects - projects_to_merge.pluck(:name)
+    if missing_projects.present?
       render_error(
-        status: 404,
-        errorcode: 'not_found',
-        message: "Project '#{@project}' doesn't have staging workflow."
+        status: 400,
+        errorcode: 'invalid_request',
+        message: "API request contains projects that are not part of the workflow: '#{missing_projects.join(', ')}'"
+      )
+      return
+    end
+    projects_not_mergable = projects_to_merge.reject { |project| project.overall_state == :acceptable }
+    if projects_not_mergable.present?
+      render_error(
+        status: 400,
+        errorcode: 'invalid_request',
+        message: "API request contains projects that are not in state acceptable: '#{projects_not_mergable.join(', ')}'"
       )
       return
     end
-    authorize staging_workflow
 
-    staging_workflow.destroy!
+    projects_to_merge.each { |project| StagingProjectMergeJob.perform_later(project.id, User.current.login) }
+    render_ok
   end
 
   private
 
+  def set_staging_projects
+    request_body = request.body.read
+    parsed_xml_hash = Xmlhash.parse(request_body) || {}
```

Comment:
please add proper xml validation and remove the || {}

the API has a schema validator built in

Reply:
@bgeuken 

you can do something like:

```
validate_action create: { method: :post, request: :staging_project }
```